File: BOJ/9373번/9373.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import kotlin.math.pow
import kotlin.math.sqrt

data class Camera(val x: Int, val y: Int, val r: Int)
data class Node(val a: Int, val b: Int, val dis: Double)

fun getDistance(a: Pair<Int, Int>, b: Pair<Int, Int>): Double {
    val xDis = (a.first - b.first).toDouble().pow(2)
    val yDis = (a.second - b.second).toDouble().pow(2)
    return sqrt(xDis + yDis)
}

lateinit var root: IntArray
fun findRoot(r: Int): Int {
    return when {
        (root[r] < 0) -> r
        else -> {
            root[r] = findRoot(root[r])
            root[r]
        }
    }
}

fun unionRoot(a: Int, b: Int): Boolean {
    val aRoot = findRoot(a)
    val bRoot = findRoot(b)
    if (aRoot == bRoot) return false
    root[bRoot] = aRoot
    return true
}

fun solution() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val t = readLine().toInt()
    repeat(t) {
        val w = readLine().toInt()
        val n = readLine().toInt()
        val cam = mutableListOf<Camera>()
        repeat(n) {
            val (x, y, r) = readLine().split(" ").map(String::toInt)
            cam.add(Camera(x, y, r))
        }
        val link = mutableListOf<Node>()
        link.add(Node(0, n + 1, w.toDouble()))
        root = IntArray(n + 2) { -1 }

        repeat(n) { index ->
            if (cam[index].x - cam[index].r <= 0) unionRoot(0, index + 1)
            else link.add(Node(0, index + 1, (cam[index].x - cam[index].r).toDouble()))
            if (cam[index].x + cam[index].r >= w) unionRoot(index + 1, n + 1)
            else link.add(Node(index + 1, n + 1, (w - cam[index].x - cam[index].r).toDouble()))

            for (next in index + 1 until n) {
                var distance = getDistance(Pair(cam[index].x, cam[index].y), Pair(cam[next].x, cam[next].y))
                if (distance <= (cam[index].r + cam[next].r)) {
                    link.add(Node(index + 1, next + 1, 0.0))
                } else {
                    distance -= (cam[index].r + cam[next].r).toDouble()
                    if (distance <= w) link.add(Node(index + 1, next + 1, distance))
                }
            }
        }
        var result = 0.0
        link.sortBy { it.dis }
        val size = link.size
        run loop@{
            repeat(size) {
                if (unionRoot(link[it].a, link[it].b)) {
                    if (findRoot(0) == findRoot(n + 1)) {
                        result = link[it].dis
                        return@loop
                    }
                }
            }
        }
        bw.write("${if (result == 0.0) 0 else String.format("%.10f", result / 2.0)}\n")
    }
    bw.close()
}

fun main() {
    solution()
}

File: BOJ/9373번/9373.cpp
#include <cstdio>
#include <vector>
#include <cmath>
#include <algorithm>
using namespace std;

struct camera {
    int x, y, r;
};
struct node {
    int a, b;
    double dis;
    bool operator<(const node &a) const {
        return dis < a.dis;
    }
};
vector<int> root;
int find_root(int a) {
    if (root[a] < 0) return a;
    return root[a] = find_root(root[a]);
}
bool union_root(int a, int b) {
    a = find_root(a);
    b = find_root(b);
    if (a == b) return false;
    root[b] = a;
    return true;
}
double get_distance(const pair<int, int> &a, const pair<int, int> &b) {
    return sqrt(pow((long long) (a.first - b.first), 2) + pow((long long) (a.second - b.second), 2));
}
int main(void) {
    int t;
    scanf("%d", &t);
    while (t--) {
        int w, n;
        scanf("%d%d", &w, &n);
        vector<camera> cam(n);
        for (int i = 0; i < n; i++) {
            scanf("%d%d%d", &cam[i].x, &cam[i].y, &cam[i].r);
        }
        vector<node> link;
        link.push_back({0, n + 1, (double) w});
        root.assign(n + 2, -1);

        for (int i = 0; i < n; i++) {
            if (cam[i].x - cam[i].r <= 0) union_root(0, i + 1);
            else link.push_back({0, i + 1, (double) (cam[i].x - cam[i].r)});
            if (cam[i].x + cam[i].r >= w) union_root(i + 1, n + 1);
            else link.push_back({i + 1, n + 1, (double) (w - cam[i].x - cam[i].r)});

            for (int j = i + 1; j < n; j++) {
                double dis = get_distance({cam[i].x, cam[i].y}, {cam[j].x, cam[j].y});
                if (dis <= (cam[i].r + cam[j].r)) {
                    link.push_back({i + 1, j + 1, 0.0});
                } else {
                    dis -= (double) (cam[i].r + cam[j].r);
                    if (dis <= w) link.push_back(node{i + 1, j + 1, dis});
                }
            }
        }
        double result = 0.0;
        sort(link.begin(), link.end());
        int l_size = link.size();
        for (int i = 0; i < l_size; i++) {
            if (union_root(link[i].a, link[i].b)) {
                if (find_root(0) == find_root(n + 1)) {
                    result = link[i].dis;
                    break;
                }
            }
        }
        if (result == 0.0) printf("0\n");
        else {
            printf("%.10lf\n", result / 2.0);
        }
    }
    return 0;
}